Add PID computation (#1314)
trixi-framework/Trixi.jl: docs/src/performance.md

- Consider using `@nospecialize` for methods like custom implementations of `Base.show`.

## Performance metrics of the `AnalysisCallback`
The [`AnalysisCallback`](@ref) computes several values that you can use to
evaluate the serial and parallel performance of Trixi. Since they represent
measured run times that are normalized by the number of `rhs!` evaluations and
the number of degrees of freedom of the problem setup, we refer to them
as "performance indices" or "PIDs". The normalization ensures that we can
compare different measurements for each type of PID independent of the number of
time steps or mesh size. All PIDs have in common that they are still in units of
time, thus *lower is better* for each of them.

Here, the term "degrees of freedom" (DOFs) refers to the number of *independent*
state vectors that are used to represent the numerical solution. For example, if
you use a DGSEM-type scheme in 2D on a mesh with 8 elements and with
5-by-5 Gauss-Lobatto nodes in each element (i.e., a polynomial degree of 4), the
total number of DOFs would be
```math
n_\text{DOFs,DGSEM} = \{number of elements\} \cdot \{number of nodes per element\} = 8 \cdot 5 \cdot 5 = 200.
```

> Similarly, for a finite volume-type schemewith 64 elements, the total number of

Reviewer: ```suggestion
Similarly, for a finite volume-type scheme with 64 elements, the total number of
```

Author: Thanks, fixed in 349b365f.

DOFs would be (independent of the number of spatial dimensions)
```math

> n_\text{DOFs,FV} = \{number of elements\} = 64,

Reviewer: ```suggestion
n_\text{DOFs,FV} = \{number of elements\} = 8,
```
You use 8 elements above?

Author: Yes, but I changed it to 64 later. Using 8 consistently makes more sense didactically, so I changed it in 349b365f.

```
since for standard finite volume methods you store a single state vector in each
element. Note that we specifically count the number of state *vectors* and not
the number of state *variables* for the DOFs. That is, in the previous example

> ``n_\text{DOFs,FV} = 64`` independent of whether this is a compressible Euler

Reviewer: ```suggestion
``n_\text{DOFs,FV} = 8`` independent of whether this is a compressible Euler
```

Author: Fixed in 349b365f.

the PIDs are repeatedly computed and can thus also be used to track the
performance over the course of a longer simulation, e.g., to analyze setups with varying performance
characteristics. Note that the time spent in the `AnalysisCallback` itself is always
*excluded*, i.e., the performance measurements are not distorted by potentially
expensive solution analysis computations. All other parts of a Trixi simulation
are included, however, thus make sure that you disable everything you do *not*
want to be measured (such as I/O callbacks, visualization etc.).

### Local, `rhs!`-only PID
The *local, `rhs!`-only PID* is computed as
```math
\text{PID}_\text{local,\texttt{rhs!}} = \frac{\{\text{accumulated time spent in \texttt{rhs!}}\}}{n_\text{DOFs,local} \cdot n_\text{calls,\texttt{rhs!}}},
```
where ``n_\text{DOFs,local}`` is the *local* number of DOFs (i.e., on the
current MPI rank; if doing a serial run, you can just think of this as *the*
number of DOFs) and ``n_\text{calls,\texttt{rhs!}}`` is the number of times the
`rhs!` function has been evaluated. Note that for this PID, we measure *only*
the time spent in `rhs!`, i.e., by definition all computations outside of `rhs!`

> - specifically all other callbacks - are not taken into account.

Reviewer: ```suggestion
- specifically all other callbacks and the time integration method - are not taken into account.
```

Author: Fixed in 349b365f.

number of DOFs) and ``n_\text{calls,\texttt{rhs!}}`` is the number of times the
`rhs!` function has been evaluated. Note that for this indicator, we measure *only*
the time spent in `rhs!`, i.e., by definition all computations outside of `rhs!` - specifically
all other callbacks and the time integration method - are not taken into account.

The local, `rhs!`-only indicator is usually most useful if you do serial
measurements and are interested in the performance of the implementation of your
core numerical methods (e.g., when doing performance tuning).

### Performance index (PID)
The *performance index* (PID) is computed as
```math
\text{PID} = \frac{\{\text{walltime since last call to \texttt{AnalysisCallback}}\} \cdot n_\text{ranks,MPI}}{n_\text{DOFs,global} \cdot n_\text{calls,\texttt{rhs!}}},
```
where ``n_\text{ranks,MPI}`` is the number of MPI ranks used,
``n_\text{DOFs,global}`` is the *global* number of DOFs (i.e., the sum of
DOFs over all MPI ranks; if doing a serial run, you can just think of this as *the*
number of DOFs) and ``n_\text{calls,\texttt{rhs!}}`` is the number of times the
`rhs!` function has been evaluated since the last call to the `AnalysisCallback`.

> 

Reviewer: ```suggestion
`rhs!` function has been evaluated since the last call to the `AnalysisCallback`.
The PID measures everything except the time spent in the `AnalysisCallback` itself -
specifically, all other callbacks and the time integration method itself are included.

```
Maybe?

In a sense, it mimics the "core hours" metric that is often used by
supercomputer centers to measure how many resources a particular compute job
requires. It can thus be seen as a proxy for "energy used" and, as an extension, "monetary cost".

!!! note "Initialization overhead in measurements"

>     When using one of `OrdinaryDiffEq`'s integration schemes, their implementation
>     will initialize some `OrdinaryDiffEq`-specific information during the first

Reviewer: ```suggestion
    When using one of the integration schemes from OrdinaryDiffEq.jl, their implementation
    will initialize some OrdinaryDiffEq.jl-specific information during the first
```

Reviewer: Just for consistency

Author: I'm loving it 😎
